packages/hebao_v2/contracts/base/libwallet/MetaTxLib.sol

```diff
@@ -154,27 +172,40 @@ library MetaTxLib
 
     function isNonceValid(
         Wallet  storage wallet,
-        uint    nonce,
-        bytes4  methodId
+        MetaTx  memory  metaTx
         )
         public
         view
         returns (bool)
     {
-        if ( methodId == SmartWallet.changeMasterCopy.selector ||
-             methodId == SmartWallet.addGuardianWA.selector ||
-             methodId == SmartWallet.removeGuardianWA.selector ||
-             methodId == SmartWallet.unlock.selector ||
-             methodId == SmartWallet.changeDailyQuotaWA.selector ||
-             methodId == SmartWallet.recover.selector ||
-             methodId == SmartWallet.addToWhitelistWA.selector ||
-             methodId == SmartWallet.transferTokenWA.selector ||
-             methodId == SmartWallet.callContractWA.selector ||
-             methodId == SmartWallet.approveTokenWA.selector ||
-             methodId == SmartWallet.approveThenCallContractWA.selector ) {
-            return nonce == 0;
-        } else {
-            return nonce > wallet.nonce && (nonce >> 128) <= block.number;
-        }
+        return (metaTx.nonce > wallet.nonce && (metaTx.nonce >> 128) <= block.number) ||
+               isDataless(metaTx);
+    }
+
+    function isDataless(
+        MetaTx memory metaTx
+        )
+        public
+        pure
+        returns (bool)
+    {
+        // We don't require any data in the meta tx when
+        // - the meta-tx has no nonce
+        // - the meta-tx needs to be successful
+        // - a function is called that requires a majority of guardians and fails when replayed
+        bytes4 methodId = metaTx.data.toBytes4(0);
+        return metaTx.nonce == 0 &&
+               metaTx.requiresSuccess &&
+               (methodId == SmartWallet.changeMasterCopy.selector ||
+                methodId == SmartWallet.addGuardianWA.selector ||
+                methodId == SmartWallet.removeGuardianWA.selector ||
```

Comment:
`resetGuardiansWA` missing from the list

Reply:
This is why I don't like these kinds of lists, so easy to forget to update them. And everything would work fine without this list...